Refactor to use bare RxJS implementations.

robwormald opened the pull request:
change RxJS implementation to use bare Observable and Subject with no included operators.

robwormald committed:
chore(package): use ultralight ObservableS

jeffbcross commented on modules/playground/src/http/http_comp.ts:
```diff
@@ -14,10 +14,8 @@ import {Http, Response} from 'angular2/http';
-    http.get('./people.json')
-        .map((res: Response) => res.json())
-        .subscribe((people: Array<Object>) => this.people = people);
+    http.get('./people.json').subscribe(res => { this.people = res.json(); });
```
Did this need to be changed as part of this PR?

robwormald replied:
Ah, I changed that when I did the first commit, I'll add it back.

jeffbcross replied:
Still needs to be fixed.

jeffbcross commented on modules/angular2/test/core/facade/observable_spec.ts:
```diff
@@ -0,0 +1,141 @@
+
+    });
+  });
+}
```
Add newline plz

jeffbcross commented on modules/angular2/test/core/facade/observable_spec.dart:
```diff
@@ -0,0 +1,5 @@
+
+main() {
+  //stub to ignore JS Observable specific tests
+}
```
Add newline plz

IgorMinar replied:
we should change clang format rather than nitpick over this.

jeffbcross replied:
I found that feature in a [fork](https://github.com/ebaker355/clang-format) of clang-format, but can't find it in clang-format itself. Looking to see if I can add a macro to clang...

jeffbcross replied:
@alexeagle or @mprobst do you have an idea on how to have clang-format enforce newlines at end of files?

jeffbcross commented on modules/angular2/src/facade/async.ts:
```diff
@@ -122,6 +131,10 @@ export class EventEmitter<T> extends Subject<T> {
+
 // todo(robwormald): ts2dart should handle this properly
 export class Observable<T> extends RxObservable<T> {
   lift<T, R>(operator: Operator<T, R>): Observable<T> {
```
What do we gain by overriding `lift`? (I know it's not part of the PR, but I think it can be removed)

robwormald replied:
it just ensures that when you use an operator, you get back an Observable of the same type as you started with. Without override, you go from someObservable (ours) => map => rxObservable. Functionally it makes no difference (unless we add any custom operators of our own, like .json() for http)

That whole override is because TS2Dart doesn't like it.

jeffbcross replied:
> That whole override is because TS2Dart doesn't like it.

TS2Dart shouldn't care about this code, since there's a separate dart implementation of async.

jeffbcross commented:
Any reason why we can't just re-export `Observable` and add to the prototype? The only failure I see is the public API increasing because of all the static method stubs on Observable.

```typescript
import {Observable} from '@reactivex/rxjs/dist/cjs/Observable';
export {Observable} from '@reactivex/rxjs/dist/cjs/Observable';

Observable.prototype.map = map;
Observable.prototype.filter = filter;
```

jeffbcross commented:
Also, @vsavkin made the valid point that we should include `flatMap` since it accepts observables, and is a good basis to chain observables. To not have `flatMap` would be crippling, and I should have added it in the original issue :(. 

I'm going to chat with @IgorMinar about this PR, because he said he has some thoughts.

jeffbcross commented:
@robwormald LGTM once you remove the last commit and finish the cleanup comments.

You could also wait until https://github.com/ReactiveX/RxJS/issues/751 and https://github.com/ReactiveX/RxJS/issues/750 land if you'd like.

IgorMinar commented on modules/angular2/src/facade/async.ts:
```diff
@@ -3,9 +3,18 @@ import {global, isPresent} from 'angular2/src/facade/lang';
 // without depending on rxjs.
 import {PromiseWrapper, Promise, PromiseCompleter} from 'angular2/src/facade/promise';
```
can you remove this unused import in a separate commit? I'd do it, but then I'd create a conflict for you

IgorMinar commented on modules/angular2/src/facade/async.ts:
```diff
@@ -122,6 +131,10 @@ export class EventEmitter<T> extends Subject<T> {
 
+// patch in Rx operators - must be done on Base class to propagate to Subject
+//RxObservable.prototype.map = map;
+//RxObservable.prototype.filter = filter;
```
this will go away, right?

it should be in rxjs not here

IgorMinar commented on modules/angular2/test/public_api_spec.ts:
```diff
@@ -706,25 +706,7 @@ var NG_ALL = [
-  'Observable#defer():js',
-  'Observable#empty():js',
-  'Observable#forkJoin():js',
-  'Observable#from():js',
```
is Observable.from going away too?

IgorMinar replied:
is that just an operator? I'm not familiar with the implementation...

robwormald replied:
yeah, all but create are static methods tacked on later

IgorMinar commented:
the rest looks good

robwormald commented:
ObservableWrapper.fromPromise / .toPromise are gonna fail now. Thoughts?

IgorMinar commented:
let's go with barebones first and then create PRs that add these things and
measure payload diff. Then decide.

On Sat, Nov 21, 2015 at 2:50 AM Rob Wormald <notifications@github.com>
wrote:

> ObservableWrapper.fromPromise / .toPromise are gonna fail now. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/5288#issuecomment-158626809>.
>

IgorMinar commented:
barebones = what's in this PR now

On Sat, Nov 21, 2015 at 2:51 AM Igor Minar <igor@angularjs.org> wrote:

> let's go with barebones first and then create PRs that add these things
> and measure payload diff. Then decide.
>
> On Sat, Nov 21, 2015 at 2:50 AM Rob Wormald <notifications@github.com>
> wrote:
>
>> ObservableWrapper.fromPromise / .toPromise are gonna fail now. Thoughts?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/angular/angular/pull/5288#issuecomment-158626809>.
>>
>

jeffbcross commented:
@robwormald is anything blocking this? If not, can you work with @alxhub to merge it?

robwormald commented:
cc @alxhub - this is merge ready

bradlygreen commented:
@robwormald CI is failing...can you take a look?

robwormald committed:
refactor(facade): use rxjs package

move to new RxJS distribution.

BREAKING CHANGE:

RxJS imports now are via `rxjs` instead of `@reactivex/rxjs`
Individual operators can be imported `import 'rxjs/operators/map'`

robwormald committed:
refactor(http): remove mockbackend export

robwormald committed:
chore(package): use rxjs package

robwormald commented:
alllllright, big changes. new version of rxjs in, minimal API surface 
- toPromise() / fromPromise() cc @vsavkin 
- take() (only in mock_backend)

@jeffbcross @IgorMinar plz discuss

robwormald commented:
will need to add following operators:
- flatMap
- map
- from (spec)
- of (spec)

pkozlowski-opensource commented on npm-shrinkwrap.clean.json:
```diff
@@ -13268,33 +13268,6 @@
         }
       }
     },
-    "source-map-loader": {
```
Why is the `source-map-loader` removed? It looks like something went wrong with the shrinkwrap process (npm install not executed?)

pkozlowski-opensource commented:
IMO shrinkwrap is broken in this PR. It needs to be corrected as it might break source maps for bundles.

jeffbcross commented on gulpfile.js:
```diff
@@ -1148,13 +1147,16 @@ var JS_DEV_DEPS = [
-  } else {
-    return source;
-  }
+  var n = source.indexOf('System.register("rxjs/Subject"');
```
We can't omit the license. What is the underlying problem?

robwormald replied:
It doesn't ship in the new repo. I can grab it from the old I guess.

jeffbcross commented on modules/angular2/src/facade/async.ts:
```diff
@@ -3,9 +3,18 @@ import {global, isPresent, noop} from 'angular2/src/facade/lang';
+import {Operator} from 'rxjs/Operator';
+
+import 'rxjs/observable/fromPromise';
+import 'rxjs/operators/toPromise';
```
Why not the basic array operators (map, filter, reduce) + flatMap?

jeffbcross commented:
TODOs:

 - [ ] Add flatMap, map, from, of, filter, reduce operators
 - [ ] Measure difference in bundle size from bare Observable and Observable + ^^^ operators
 - [ ] Include rxjs in the benchpress bundle (it is required, and the previous exclude pattern doesn't seem to have actually been working)
 - [ ] Fix shrinkwrap to not remove sourcemap
 - [ ] Remove PromiseWrapper unused import in a separate commit
 - [ ] Restore E2E test to its former Rx-ified glory
 - [ ] Find a way to include the RxJS license

Regarding `take` in mock_backend, as long as we only import it in a testing barrel, no need to be concerned with added bytes.

I'll take care of all of the todos except the last one for now. @robwormald do you want to try to correct the license issue upstream and convince @blesh to cut another RxJS release?

mprobst commented:
@jeffbcross I filed an issue for the newline-at-EOF, I don't think that's supported yet.

jeffbcross commented:
Thanks @mprobst!

blesh commented:
> Find a way to include the RxJS license

Is this to say that the new packages are missing the LICENSE file?... that's not good. lol

> I'll take care of all of the todos except the last one for now. @robwormald do you want to try to correct the license issue upstream and convince @blesh to cut another RxJS release?

The beta release is coming this week (hopefully in a couple of days). Does it need to be before that?

robwormald commented:
I'll fix the Rx stuff over there first, and in the interim can just grab
the license from the @reactivex package.
On Tue, Dec 1, 2015 at 12:03 PM Ben Lesh <notifications@github.com> wrote:

> Find a way to include the RxJS license
>
> Is this to say that the new packages are missing the LICENSE file?...
> that's not good. lol
>
> I'll take care of all of the todos except the last one for now.
> @robwormald <https://github.com/robwormald> do you want to try to correct
> the license issue upstream and convince @blesh <https://github.com/blesh>
> to cut another RxJS release?
>
> The beta release is coming this week (hopefully in a couple of days). Does
> it need to be before that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/5288#issuecomment-161079651>.
>

jeffbcross commented:
Bundle size measurements (using the reported numbers from `gulp bundles.js.checksize` for gzip level 2 as well as running `gzip --best -c dist/build/angular2.min.js > dist/build/angular2.min.js.gz` for gzip level 9).

| variant | gzip level 2 (bytes) | gzip level 9 (bytes) |
|-----------|----------------|-------------------|
| with no operators | 153,670  | 118,022 |
| with 8 operators* | 160,233 | 123,755 |
| **difference from bare** | -5563 | -5733 |
| with 4 operators** | 156303 | 120,250 |
| **difference from bare** | -2633 | -2228 |

*these are the operators included in the "8 operators"
```typescript
import 'rxjs/operators/map';
import 'rxjs/operators/mergeMap'; //aka flatMap
import 'rxjs/operators/filter';
import 'rxjs/operators/reduce';
import 'rxjs/operators/toPromise';

import 'rxjs/observable/from';
import 'rxjs/observable/fromArray'; //Observable.of
import 'rxjs/observable/fromPromise';
```

**these are the operators included in the "4 operators"
```typescript
import 'rxjs/operators/map';
import 'rxjs/operators/mergeMap'; //aka flatMap
import 'rxjs/operators/filter';
import 'rxjs/operators/reduce';
```

CC @IgorMinar

blesh commented:
FWIW, `Observable.of` and `Observable.from` are part of the Observable spec.

jeffbcross commented:
FYI I opened https://github.com/angular/angular/pull/5533 to fix some of the remaining issues here (everything except the license issue)
